testv2/e2e/cloudprovider.go

```diff
@@ -0,0 +1,381 @@
+/*
+Copyright 2022 The KubeOne Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func supportsCloudProviderTests(provider string) bool {
+	switch provider {
+	case "aws", "digitalocean", "hetzner", "gce", "openstack":
+		return true
+	default:
+		return false
+	}
+}
+
+const (
+	cpTestNamespaceName   = "test-cloud-provider"
+	cpTestStatefulSetName = "test-sts"
+	cpTestServiceName     = "test-svc"
+
+	cpTestPollPeriod = 10 * time.Second
+	cpTestTimeout    = 10 * time.Minute
+)
+
+var (
+	cloudProviderPodLabels = map[string]string{"app": "test-cp"}
+)
+
+type cloudProviderTests struct {
+	ctx      context.Context
+	client   ctrlruntimeclient.Client
+	provider string
+}
+
+func newCloudProviderTests(client ctrlruntimeclient.Client, provider string) *cloudProviderTests {
+	return &cloudProviderTests{
+		ctx:      context.Background(),
+		client:   client,
+		provider: provider,
+	}
+}
+
+func (c *cloudProviderTests) run(t *testing.T) {
+	if !supportsCloudProviderTests(c.provider) {
+		t.Logf("Skipping cloud provider tests because cloud provider %q is not supported.", c.provider)
+
+		return
+	}
+
+	c.createStatefulSetWithStorage(t)
+	c.exposeStatefulSet(t)
+}
+
+func (c *cloudProviderTests) runWithCleanup(t *testing.T) {
```

Comment:
Why not cleanUp always inside run()? If we didn't created anything, cleanUp() would simply return after few 404 GETs.

Reply:
Please see https://github.com/kubermatic/kubeone/pull/2177#discussion_r940343385